### ISSUE TITLE
cmd/link: PIC support for darwin/arm

### ISSUE THREAD

**Issue by KennyAtBaidu** [golang/go#12896]
When including the generated gomobile iOS framework into another iOS framework, calling gomobile generated method from the wrapper framework causes linker failure with message "ld: illegal text-relocation to 'type..eq.[0]string'"

**Comment by rakyll** [comment]
Duplicate of https://github.com/golang/go/issues/9210.

**Comment by rakyll** [comment]
The only workaround before #9210 is fixed is to keep targeting SDK version 22.

```
<uses-sdk
        android:minSdkVersion="15"
        android:targetSdkVersion="22" />
```

**Comment by KennyAtBaidu** [comment]
@rakyll Sorry if I wasn't clear. This is for iOS framework. When code of custom iOS framework calling methods that are generated by gomobile cause illegal text-relocation.

**Comment by rakyll** [comment]
Sorry, it was not clear from the text.

/cc @crawshaw

**Comment by KennyAtBaidu** [comment]
updated the original report.

**Comment by crawshaw** [comment]
Hi @KennyAtBaidu. Can you tell if this is coming from the ARMv7 or ARMv8 (arm64) binary? (You can extract the individual .a files by running gomobile with the -work.)

For ARMv7 it looks like we are not compiling as PIC for iOS, which is easy to fix. For ARMv8 I was under the impression the compiler was generating PC-relative code by default, but I tried to confirm with a quick inspection of the compiler and it's not clear to me.

**Comment by KennyAtBaidu** [comment]
Hi @crawshaw 
It seems like 
```
bind -work 
```
gives me x86_64, armv7, and arm64 architecture. I'm not really familiar about PIC, is there anything I can do on my end?

**Comment by crawshaw** [comment]
Ok great, then this might be easy to fix. I'll try to attach a cl later today.

**Comment by hsinhoyeh** [comment]
do we have any updates? I also hit into this.
can I know how to workaround this if we cann't fix it immediately.

**Comment by clsung** [comment]
same here. And thank you for your great effort.

**Comment by crawshaw** [comment]
I spent a couple of unproductive days trying to make PC-relative relocations work with darwin/arm. I'm close: the right instructions are generated, it's just the arithmetic is off. It will probably take me the rest of the week to make this work, assuming no other serious bugs appear.

In the mean time, you can extract the arm64 binaries as described above and work exclusively with those. This means your programs won't run on older iOS hardware, but it may be enough to keep you going for now.

**Comment by crawshaw** [comment]
I have not had any luck generating a pc-relative macho relocation on arm. Every relocation I've tried (including SECTDIFF encoded as a scattered relocation) has ignored the PC-relative bit, and instead gives me an absolute relocation or zero.

If anyone has any ideas I'm all ears.

Meanwhile, next week I'll try setting up the exact environment that produced this error and see what about the text relocation is bothering it. Maybe there's some other way to make Xcode happy.

**Comment by damouse** [comment]
I can see the individual statics with `-work`, can you briefly explain how I can get them back into xcode?

Otherwise is there another quick way of getting the embedded framework to function?

**Comment by dcu** [comment]
I just tested this and it works fine (except for a warning) with armv7+arm64. It stops working when I add amd64 to the fat binary.

**Comment by dcu** [comment]
I did something like this to generate the binary without amd64:

```bash
FRAMEWORK=Mysdk # this is your package name capitalized
gomobile bind -target=ios -prefix=$FRAMEWORK
cd $FRAMEWORK.framework/Versions/A
lipo Coresdk -thin armv7 -output /tmp/armv7.a
lipo Coresdk -thin arm64 -output /tmp/arm64.a
lipo -create /tmp/armv7.a /tmp/arm64.a -output $FRAMEWORK
```

**Comment by dcu** [comment]
@crawshaw hi, is there anything we can help with to help you fix this issue?

**Comment by dcu** [comment]
I backported https://github.com/golang/go/commit/aea4222f673ee9189ba81548978b091004afe994 to go 1.6 and it's working way better.

**Comment by dcu** [comment]
@crawshaw @cherrymui thanks for your hard work on this issue.

I tested this on tip but I'm getting the following error:
```shell
gomobile bind -target=ios
gomobile: reading export data: /var/folders/y_/ws09jpg90_b5y710wmd7kvj5sbl416/T/gomobile-work-499035079/fakegopath/pkg/darwin_arm/path/to/my/sdk.a: unknown version:
make: *** [ios] Error 1
```

so I backported the commit to go 1.6 but when compiling I'm getting a similar error to the previous one:

```shell
ld: illegal text-relocation to 'go.string.*' in /test/sdk.framework/sdk(go.o) from 'runtime.interhash' in /test/sdk.framework/sdk(go.o) for architecture armv7
```

Let me know if we can provide more information.

**Comment by crawshaw** [comment]
@dcu is the first error in your comment, about "reading export data", fixed? I believe I saw the CL go by updating the gcimporter, and have no been able to replicate that error message.

**Comment by dcu** [comment]
@crawshaw no, sorry I forgot to update here. 

This got fixed after #15519 and updating gomobile.  it works on both 1.6 and 1.7.
Now we are getting errors when trying to install on a armv7 device because of a signature (according to the error message). Some of the issues are reported here: #15657 
The other issue is that enabling armv7 requires the 386 sim I'm not sure why.
